platform_generator.R

```diff
@@ -156,68 +107,62 @@ for(j in 1:nrow(gate_db)){
                                          &Download_Start <= (step*(s-1)+gate_downloads[i,]$Download_Start)
                                          &Download_End >= (step*(s-1)+gate_downloads[i,]$Download_Start)))
   }
-  # take the mean of concurrency of all sampling points
-  if(mean(tmp) > 1){
-    gate_downloads[i,]$concurrency <- mean(tmp)
-  }
-}
 
-## function to find the concurrency of a link
-## if the max is given by several transfers, the maximum concurrency is taken for this link.
-Concurrency_link <- function(Bandwidth, concurrency){
-  pos <- which(Bandwidth == max(Bandwidth))
-  con <- concurrency[pos[1]]
-  if(length(pos)>1){
-    for(i in 2:length(pos)){
-      if(concurrency[pos[i]]> con){
-        con <- concurrency[pos[i]]
-      }
-    }
-  }
-  return (con)
+  tmp <- 1/conc
+  gate_downloads[i,]$concurrency <- n_interval/sum(tmp)
 }
 
+gate_downloads$Bandwidth_improved <- gate_downloads$Bandwidth * gate_downloads$concurrency
+merge_downloads$Bandwidth_improved <- merge_downloads$Bandwidth * merge_downloads$concurrency
+
+
+df_download <- rbind(gate_downloads, merge_downloads)
 
-Links = ddply(subset(gate_downloads,FileSize == max(gate_downloads$FileSize)),
-           c("SE_SITE"), summarize, 
-           Ratio=max(Bandwidth)/mean(Bandwidth),
-           concurrency= Concurrency_link(Bandwidth,concurrency),
-           .drop=TRUE)
 
+##############################
+
+# Compute the respective average bandwidth from (then to) these SE to (then from)each grid site. Let ddply produce NaN
+# entries. The rationale is that during the simulation the LFC can pick a SE for download input that was not selected 
+# during the real execution. To circumvent this, we add a default bandwidth for the missing connections. This value 
+# is set to the maximum observed bandwidth.
 
-Target_links <- subset(Links, Ratio <= 1.696)
+SE_to_site_bandwidth = ddply(df_download, c("Source","SiteName"), summarize, 
+                             AvgBandwidth=round(mean(Bandwidth),2), MaxBandwidth=max(Bandwidth_improved), .drop=FALSE)
 
-# apply the factor to all target links
-if(nrow(Target_links) !=0){
-  for(i in 1:nrow(Target_links)){
 
-    SE_to_site_bandwidth[SE_to_site_bandwidth$SE_SITE ==Target_links[i,]$SE_SITE,]$MaxBandwidth<- 
-    SE_to_site_bandwidth[SE_to_site_bandwidth$SE_SITE ==Target_links[i,]$SE_SITE,]$MaxBandwidth*Target_links[i,]$concurrency
+if (TRUE %in% is.nan(SE_to_site_bandwidth$AvgBandwidth)) {
+  SE_to_site_bandwidth[is.nan(SE_to_site_bandwidth$AvgBandwidth),]$AvgBandwidth <- 
+    max(transfers[transfers$UpDown==2,]$Bandwidth)
+}
 
-  } 
+if (TRUE %in% is.infinite(SE_to_site_bandwidth$MaxBandwidth)) {
+  SE_to_site_bandwidth[is.infinite(SE_to_site_bandwidth$MaxBandwidth),]$MaxBandwidth <- 
+    max(transfers[transfers$UpDown==2,]$Bandwidth)
 }
 
-##############################
+SE_to_site_bandwidth$SE_SITE <- paste(SE_to_site_bandwidth$Source,'_',SE_to_site_bandwidth$SiteName,sep='')
```

Comment:
Its the third time we build the SE_SITE field the same way. This has to be done at the beginning for the "transfers" data frame (around line 50)